src/Server/SSH/SSHPtyHandler.cpp

```diff
@@ -0,0 +1,511 @@
+#include <atomic>
+#include <stdexcept>
+#include <Server/EmbeddedClient/openpty.h>
+#include <Common/SSH/clibssh.h>
+#include <Server/SSH/SSHPtyHandler.h>
+#include <boost/iostreams/device/file_descriptor.hpp>
+#include <boost/iostreams/stream.hpp>
+#include <sys/poll.h>
+#include <Poco/Net/StreamSocket.h>
+#include <Poco/Pipe.h>
+#include "Access/Common/AuthenticationType.h"
+#include "Access/Credentials.h"
+#include "Access/SSH/SSHPublicKey.h"
+#include "Core/Names.h"
+#include "Server/EmbeddedClient/EmbeddedClientRunner.h"
+#include "Server/EmbeddedClient/IClientDescriptorSet.h"
+#include "Server/EmbeddedClient/PtyClientDescriptorSet.h"
+#include "Server/SSH/SSHChannel.h"
+#include "Server/SSH/SSHEvent.h"
+#include <Common/logger_useful.h>
+
+namespace
+{
+
+/*
+Need to generate adapter functions, such for each member function, for example:
+
+class SessionCallback
+{
+For this:
+    ssh_channel channelOpen(ssh_session session)
+    {
+        channel = SSHChannel(session);
+        return channel->get();
+    }
+
+
+Generate this:
+    static ssh_channel channelOpenAdapter(ssh_session session, void * userdata)
+    {
+        auto * self = static_cast<SessionCallback*>(userdata);
+        return self->channel_open;
+    }
+}
+
+We just static cast userdata to our class and then call member function.
+This is needed to use c++ classes in libssh callbacks.
+Maybe there is a better way? Or just write boilerplate code and avoid macros?
```

Comment:
I don't get why we don't write the individual handlers (everything below l. 67)  in the first place in "adapter form".

Reply:
I wanted to use c++ methods as is and just passing them to c library. Basically I wanted to avoid writing c code as much as possible

Reply:
I wanted to use c++ methods as is and just passing them to c library. Basically I wanted to avoid writing c code as much as possible